rbd: use DeepCopy to create a thick-provisioned clone

nixpanic opened the pull request:
To create a full-allocated RBD image from a snapshot/clone DeepCopy()
can be used. This is needed when the parent of the new volume is
thick-provisioner, so that the new volume is independent of the parent
and thick-provisioned as well.

Things that need to be done:

- [x] manually tested and confirmed allocations with `rbd du`
- [x] rebase on top of the next go-ceph release (v0.10.0 #2147)
- [x] add e2e verification test

Fixes: #2077

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

nixpanic committed:
cleanup: move snapshot/clone/flatten into its own function

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
rbd: use DeepCopy to create a thick-provisioned clone

To create a full-allocated RBD image from a snapshot/clone DeepCopy()
can be used. This is needed when the parent of the new volume is
thick-provisioner, so that the new volume is independent of the parent
and thick-provisioned as well.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
e2e: validate thick-provisioned PVC-PVC cloning

Add a case to create a new PVC with VolumeContentSource set to a
thick-provisioned PVC. This should result in a new thick-provisioned PVC
once the cloning is done.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic commented:
/retest ci/centos/mini-e2e-helm/k8s-1.19

nixpanic commented:
This PR depends on #2147, once that is merged, the **[DNM]** label can be dropped.

nixpanic commented:
@mergifyio rebase

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -1505,6 +1505,29 @@ func (rv *rbdVolume) RepairThickProvision() error {
+		return err
+	}
+
+	image, err := rv.open()
```
image needs to be closed?

nixpanic replied:
yes! will add a defer for it.

Madhu-1 commented:
tested it locally its failed to mount the clone PVC to application pod
```log
Events:
  Type     Reason                  Age                   From                     Message
  ----     ------                  ----                  ----                     -------
  Normal   Scheduled               5m29s                 default-scheduler        Successfully assigned default/csirbd-demo-pod to minicluster2
  Normal   SuccessfulAttachVolume  5m29s                 attachdetach-controller  AttachVolume.Attach succeeded for volume "pvc-16c4ed55-9f7a-487a-8040-caf9c56475f2"
  Warning  FailedMount             3m26s                 kubelet                  Unable to attach or mount volumes: unmounted volumes=[mypvc], unattached volumes=[mypvc default-token-glr45]: timed out waiting for the condition
  Warning  FailedMount             72s                   kubelet                  Unable to attach or mount volumes: unmounted volumes=[mypvc], unattached volumes=[default-token-glr45 mypvc]: timed out waiting for the condition
  Warning  FailedMount             57s (x10 over 5m13s)  kubelet                  MountVolume.MountDevice failed for volume "pvc-16c4ed55-9f7a-487a-8040-caf9c56475f2" : rpc error: code = Internal desc = an error (exit status 2) occurred while running ceph args: [rbd task add flatten replicapool/csi-vol-fdb11c26-cd8f-11eb-836d-0242ac110005 --id csi-rbd-node --keyfile=***stripped*** -m 192.168.50.61:6789]
```

Madhu-1 commented:
When No PVC is present on the kubernetes the images exists in the trash
and the ceph cluster is still using 16 Gib (PVC size was 15 Gib)

```
[root@dhcp53-170 rbd]# kubectl delete -f pvc-clone.yaml 
persistentvolumeclaim "rbd-pvc-clone" deleted
[root@dhcp53-170 rbd]# kubectl delete -f pvc.yaml 
persistentvolumeclaim "rbd-pvc" deleted
[root@dhcp53-170 rbd]# kubectl get pvc
No resources found in default namespace.
[root@dhcp53-170 rbd]# kubectl get pv
No resources found
```
```
# ceph -s
  cluster:
    id:     d6d71dda-7470-418a-9186-e63e55f71efc
    health: HEALTH_WARN
            mon is allowing insecure global_id reclaim
 
  services:
    mon: 1 daemons, quorum a (age 18m)
    mgr: a(active, since 18m)
    osd: 1 osds: 1 up (since 18m), 1 in (since 18m)
 
  data:
    pools:   2 pools, 33 pgs
    objects: 3.85k objects, 15 GiB
    usage:   16 GiB used, 34 GiB / 50 GiB avail
    pgs:     33 active+clean
 
  io:
    client:   681 B/s rd, 0 op/s rd, 0 op/s wr
 
sh-4.4# rbd ls --pool=replicapool
sh-4.4# 
sh-4.4# rbd trash ls --pool=replicapool
10be52dc3ad6 csi-vol-fdb11c26-cd8f-11eb-836d-0242ac110005-temp
10be534682a8 csi-vol-d6a94576-cd8f-11eb-836d-0242ac110005
```

Madhu-1 commented:
when provisioner pod is restarted when Thick clone is getting created the thick clone will not consume expected size on ceph cluster

```
[root@dhcp53-170 rbd]# kubectl get pvc
NAME            STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS      AGE
rbd-pvc         Bound    pvc-6ce1452f-689d-4903-a6b4-af4720151818   15Gi       RWO            rook-ceph-block   69s
rbd-pvc-clone   Bound    pvc-caff5817-86a2-45df-843f-19ad8cd81e8a   15Gi       RWO            rook-ceph-block   37s

sh-4.4# rbd du replicapool/csi-vol-d791f254-cd92-11eb-8352-0242ac110005
warning: fast-diff map is not enabled for csi-vol-d791f254-cd92-11eb-8352-0242ac110005. operation may be slow.
NAME                                          PROVISIONED  USED   
csi-vol-d791f254-cd92-11eb-8352-0242ac110005       15 GiB  936 MiB
sh-4.4# rbd du replicapool/csi-vol-c477b45f-cd92-11eb-8352-0242ac110005
warning: fast-diff map is not enabled for csi-vol-c477b45f-cd92-11eb-8352-0242ac110005. operation may be slow.
NAME                                          PROVISIONED  USED  
csi-vol-c477b45f-cd92-11eb-8352-0242ac110005       15 GiB  15 GiB
sh-4.4# rbd info replicapool/csi-vol-c477b45f-cd92-11eb-8352-0242ac110005
rbd image 'csi-vol-c477b45f-cd92-11eb-8352-0242ac110005':
	size 15 GiB in 3840 objects
	order 22 (4 MiB objects)
	snapshot_count: 1
	id: 124edb32d808
	block_name_prefix: rbd_data.124edb32d808
	format: 2
	features: layering, operations
	op_features: clone-parent, snap-trash
	flags: 
	create_timestamp: Tue Jun 15 04:33:03 2021
	access_timestamp: Tue Jun 15 04:33:03 2021
	modify_timestamp: Tue Jun 15 04:33:03 2021
sh-4.4# rbd info replicapool/csi-vol-d791f254-cd92-11eb-8352-0242ac110005
rbd image 'csi-vol-d791f254-cd92-11eb-8352-0242ac110005':
	size 15 GiB in 3840 objects
	order 22 (4 MiB objects)
	snapshot_count: 0
	id: 124e9cae7648
	block_name_prefix: rbd_data.124e9cae7648
	format: 2
	features: layering, deep-flatten
	op_features: 
	flags: 
	create_timestamp: Tue Jun 15 04:33:33 2021
	access_timestamp: Tue Jun 15 04:33:33 2021
	modify_timestamp: Tue Jun 15 04:33:33 2021
sh-4.4# 

```

Madhu-1 commented on internal/rbd/clone.go:
```diff
@@ -147,6 +146,59 @@ func (rv *rbdVolume) generateTempClone() *rbdVolume {
+	}
+	defer j.Destroy()
+
+	// TODO: if rv exists, delete the image and start over?
```
if rv exists the call won't come here as the image will be found  at https://github.com/ceph/ceph-csi/blob/devel/internal/rbd/rbd_journal.go#L283 ?

nixpanic replied:
Yes, probably. The image metadata should have the thick-provisioned key, if it is missing copying the image should start over again.

nixpanic committed:
rbd: no need to flatten thick-provisioned images

Thick-provisioned images are independent, cloned images or snapshots are
deep-flattened during creation. There is no need to try and flatten them
again.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic commented:
> when provisioner pod is restarted when Thick clone is getting created the thick clone will not consume expected size on ceph cluster

In this case, `DeepCopy()` was aborted and needs to be restarted. This is still TODO at the moment.

Madhu-1 commented:
> > when provisioner pod is restarted when Thick clone is getting created the thick clone will not consume expected size on ceph cluster
> 
> In this case, `DeepCopy()` was aborted and needs to be restarted. This is still TODO at the moment.

:+1: This will be addressed in the next PR?

humblec commented on internal/rbd/clone.go:
```diff
@@ -197,41 +236,39 @@ func (rv *rbdVolume) createCloneFromImage(ctx context.Context, parentVol *rbdVol
+	} else {
+		// flatten clone
+		errFlatten = tempClone.flattenRbdImage(ctx, rv.conn.Creds, false, rbdHardMaxCloneDepth, rbdSoftMaxCloneDepth)
+		if errFlatten != nil {
```
@nixpanic if `Flatten` failed here, do we need to return error ? or can we move ahead without failing? I understand,  this is normal/general workflow we have so far/till now, so ideally should keep the same. But double confirming, is it mandatory to flatten before creating a snap from temporary clone image?.

nixpanic commented:
On Wed, Jun 16, 2021 at 05:19:42AM -0700, Humble Devassy Chirammal wrote:
> @humblec commented on this pull request.
> 
> 
> 
> >  		if err != nil {
> -			return fmt.Errorf("failed to copy encryption config for %q: %w", rv, err)
> +			return fmt.Errorf("failed to deep copy %q into %q: %w", parentVol, rv, err)
> +		}
> +	} else {
> +		// flatten clone
> +		errFlatten = tempClone.flattenRbdImage(ctx, rv.conn.Creds, false, rbdHardMaxCloneDepth, rbdSoftMaxCloneDepth)
> +		if errFlatten != nil {
> 
> @nixpanic if `Flatten` failed here, do we need to return error ? or can we move ahead without failing? I understand,  this is normal/general workflow we have so far/till now, so ideally should keep the same. But double confirming, is it mandatory to flatten before creating a snap from temporary clone image?. 

Sorry, I am not sure what you are asking...

The previous behaviour is not changed for non-thick-provisioned cloning.
Is that something you are concerned about?

For thick-provisioned cloning, the DeepCopy() function is expected to do
the flattening. I might have understood that incorrectly, as it seems
the 'deep-flatten' option is still enabled once DeepCopy() has finished.
In that case, additional flattening might still be required?

nixpanic committed:
rbd: disable FeatureDeepFlatten when doing DeepCopy()

Not all Linux kernels support the deep-flatten feature. Disabling the
feature makes it possible to map RBD images on older kernels (like what
minikube uses).

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
rbd: restart thick-provisioned PVC cloning after aborting

In case cloning a thick-PVC failed during DeepCopy(), the image will
exist, but have partial contents. Only when the image has the
thick-provisioned metadata set, it has completed DeepCopy().

When the metadata is missing, the image is deleted, and an error is
returned to the caller. Kubernetes will automatically retry provisioning
on the ABORTED error, and the cloning will get restarted from the
beginning.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic commented:
> > > when provisioner pod is restarted when Thick clone is getting created the thick clone will not consume expected size on ceph cluster
> > 
> > 
> > In this case, `DeepCopy()` was aborted and needs to be restarted. This is still TODO at the moment.
> 
> +1 This will be addressed in the next PR?

I am including that in this PR, with an additional commit.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -366,7 +366,23 @@ func (cs *ControllerServer) repairExistingVolume(ctx context.Context, req *csi.C
+					return nil, status.Errorf(codes.Aborted, "failed to delete partially cloned volume %q: %s", rbdVol, err)
+				}
+				return nil, status.Errorf(codes.Aborted, "cloning thick-provisioned volume %q has been interrupted, please retry", rbdVol)
+			}
```
if the image is deleted when the next request comes the omap will get deleted. do we need to wait for the next CreateVolume to handle omap deletion or we can delete the reserved omap entry here only?

nixpanic replied:
While testing I noticed that this isn't sufficient, the image entry needs to be removed from the journal here too.

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -1508,6 +1508,49 @@ func (rv *rbdVolume) RepairThickProvision() error {
+	}
+
+	// deep-flatten is not supported by all clients, so disable it
+	return dest.DisableDeepFlatten()
```
Question: instead of disabling for all clients. can we disable this during mapping time in node server https://github.com/ceph/ceph-csi/blob/devel/internal/rbd/nodeserver.go#L270?

nixpanic replied:
I prefer to have a single approach everywhere. Having deep-flatten enabled on some images and not on others sounds confusing.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -366,7 +366,23 @@ func (cs *ControllerServer) repairExistingVolume(ctx context.Context, req *csi.C
+			if err != nil {
+				return nil, status.Errorf(codes.Aborted, "failed to verify thick-provisioned volume %q: %s", rbdVol, err)
+			} else if !thick {
+				err = deleteImage(ctx, rbdVol, cr)
```
@nixpanic is this manually tested by restarting the provisioner pod and working fine?

nixpanic replied:
manually testing it, but it is not working completely yet

Madhu-1 reviewed (commented):
Left few questions LGTM.

nixpanic commented:
Please review again :pray: Manual tests now succeed (including provisioner restart while cloning).

humblec commented on e2e/rbd_helper.go:
```diff
@@ -424,6 +430,16 @@ func validateEncryptedPVC(f *framework.Framework, pvc *v1.PersistentVolumeClaim,
 	return validateEncryptedImage(f, rbdImageSpec, app)
 }
 
+func isThickPVC(f *framework.Framework, pvc *v1.PersistentVolumeClaim, app *v1.Pod) error {
```
@nixpanic may be a quick source comment about these functions `isThickPVC and isEncryptedPVC`  can be added

nixpanic replied:
I do not think that is needed, the names of the functions are pretty self explaining.

humblec commented:
@nixpanic LGTM, a minor comment from me, but that can be ignored too.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -366,7 +366,27 @@ func (cs *ControllerServer) repairExistingVolume(ctx context.Context, req *csi.C
+			if err != nil {
+				return nil, status.Errorf(codes.Aborted, "failed to verify thick-provisioned volume %q: %s", rbdVol, err)
+			} else if !thick {
+				err = deleteImage(ctx, rbdVol, cr)
```
still, need to delete the intermediate clone and temp snapshot?

nixpanic replied:
yes, I think so...

nixpanic replied:
using `cleanUpSnapshot()` instead of `deleteImage()` should take care of this, I think.

humblec commented on internal/rbd/rbd_util.go:
```diff
@@ -1508,6 +1508,49 @@ func (rv *rbdVolume) RepairThickProvision() error {
+	opts := librbd.NewRbdImageOptions()
+	defer opts.Destroy()
+
+	// when doing DeepCopy, also flatten the new image
```
@nixpanic @Madhu-1 I am not sure we have to explictly do  a flattening of the new image here . iic,after deep copy the new image is flattened,in that case we dont need to do that, but I could be wrong

nixpanic replied:
That is done by passing `ImageOptionFlatten` which is done on the next line, hence this comment.
